### PR TITLE
Actions

### PR DESCRIPTION
**Pull request by melchor629**
Adds implementation of the **actions** !!

Actions are a unit of work that does something inside a task. A pipeline is conformed with actions that are run at the same time passing data from one to another. The implementation implements all actions defined in the `utils.sh` file, plus some others. Documentation was also modified to reflect the changes.

## Testing

To test actions, create a file in the repo (the name does not matter) with the following contents:

```python
import logging

from mdbackup.actions.builtin._register import register
from mdbackup.actions.runner import run_task_actions


logging.basicConfig(level=logging.DEBUG)


def main():
    register()
    prev_backup_path = None  # If desired, path of the previous backup
    backup_path = '.'  # Path where the files will be stored
    actions = [
        # TO BE FILLED
    ]

    for it in actions:
        for _, value in it.items():
            if isinstance(value, dict):
                value['_backup_path'] = backup_path
                value['_prev_backup_path'] = prev_backup_path

    run_task_actions('action_test', actions)


if __name__ == '__main__':
    main()
```

Fill the `actions` with the configuration for each action in a dict. Then run the script as `PYTHONPATH=. python path/to/the/script.py` (inside the virtual env).

Example of actions configuration:

```python
actions = [
  {'asuswrt': {'host': '192.168.2.1', 'user': 'admin', 'password': 'password', 'backupType': 'jffs'}},
  {'compress-gzip': {}},
  {'to-file': {'path': 'jffs.tar.gz'}},
]
```

### CODE REVIEW

File: mdbackup/actions/builtin/command.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+import shlex
+import subprocess
+
+from mdbackup.actions.container import action, unaction
+
+
+_logger = logging.getLogger(__name__)
+
+
+def _parse_command(command: str) -> str:
```
Comment (melchor629) [inline mdbackup/actions/builtin/command.py]: The return type should be `List[str]`

File: mdbackup/actions/builtin/command.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+    if workdir is not None:
+        args.extend(['-w', workdir])
+
+    if pull is not None:
+        _logger.info(f'Pulling image {image}...')
+        proc = action_command(None, {'command': f'docker image pull -q "{image}"'})
+        stdout, stderr = proc.communicate()
+        if proc.returncode != 0:
+            _logger.error(f'Image pull for {image} failed:\n{stderr.decode("utf-8")}')
```
Comment (melchor629) [inline mdbackup/actions/builtin/command.py]: Should raise an exception to end the action

File: mdbackup/actions/builtin/database.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    password = params.get('password')
+    database = params['database']
+    port = params.get('port')
+
+    args = ['mysqldump', '-h', host]
+    if user is not None:
+        args.extend(['-u', user])
+    if password is not None:
+        params['env'] = {'MYSQL_PWD': password}
+        # args.append(f'-p{password}')
```
Comment (melchor629) [inline mdbackup/actions/builtin/database.py]: Remove commented code pls :(

File: mdbackup/actions/builtin/directory.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+                                       stream=entry.stream,
+                                       link_content=entry.link_content),
+                _recurse_dir(resolved, resolved, True),
+            )
+        else:
+            yield DirEntry.from_real_path(path, root_path)
+    elif path.is_file():
+        yield DirEntry.from_real_path(path, root_path)
+    # Ignore the rest of file types
+    # TODO check if . is included among the results
```
Comment (melchor629) [inline mdbackup/actions/builtin/directory.py]: Resolve this comment to accept the PR

File: mdbackup/actions/builtin/network.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        'Host': host,
+        'User-Agent': 'mdbackup via requests (python)',
+        'Referer': f'http://{host}/Main_Login.asp',
+    }
+    res = s.get(f'http://{host}/login.cgi', data=login_data, headers=headers)
+
+    if res.status_code != 200:
+        raise PermissionError(f'[{res.status_code}] Authentication failed: {res.text}')
+
+    headers['Referer'] = 'http://192.168.2.1/Advanced_SettingBackup_Content.asp'
```
Comment (melchor629) [inline mdbackup/actions/builtin/network.py]: Hey that IP should be converted into `{host}` :)

File: mdbackup/actions/container.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    action1 = _actions[id1]
+    action2 = _actions[id2]
+
+    if action1.output is None:
+        return f'{id1} cannot be connected to {id2}: {id1} has no output'
+    elif action1.output.startswith('stream'):
+        if action2.input != 'stream':
+            return f'{id1} cannot be connected to {id2}: {id2} expected input is not a stream'
+    else:
+        # if action1.output.startswith('directory'):
```
Comment (melchor629) [inline mdbackup/actions/container.py]: Remove the commented code

**Comment by melchor629** [comment]
Actions to be tested/already tested:

- [x] tar
- [x] command
- [x] ssh
  - [x] Using password (`sshpass`)
  - [x] Using identity file
- [x] docker
- [x] compress-xz
- [x] compress-gzip
- [x] compress-bz2
- [x] compress-br
- [X] postgres-database
  - [X] Using docker
  - [X] Not using docker
- [ ] mysql-database
  - [ ] Using docker
  - [ ] Not using docker
- [x] influxdb
  - [x] Using docker
  - [x] Not using docker
- [x] from-file
- [x] from-file-ssh
  - [x] Using password (`sshpass`)
  - [x] Using identity file
- [x] to-file
- [x] copy-file
- [x] clone-file
  - [x] With `reflink` set and in a `btrfs` fs
  - [x] With `reflink` set and not in a `btrfs` fs
  - [x] Using hardlink
- [x] from-directory
- [x] from-physical-docker-volume
- [x] to-directory
- [x] copy-directory
- [x] encrypt-gpg
- [x] mikrotik
- [X] asuswrt

**Review by amgxv — APPROVED** [review]

**Pull request merged**